net/respondd/src/respondd.c

```diff
@@ -361,13 +417,132 @@ static void serve(int sock) {
 	}
 
 	if (output) {
-		if (sendto(sock, output, output_bytes, 0, (struct sockaddr *)&addr, addrlen) < 0)
+		if (sendto(sock, output, output_bytes, 0, addr, sizeof(struct sockaddr_in6)) < 0)
 			perror("sendto failed");
 	}
 
 	json_object_put(result);
 }
 
+void serve_request(struct request_task *task, int sock) {
+	bool compress;
+	struct json_object *result = handle_request(task->request, &compress);
+
+	if (!result)
+		return;
+
+	send_response(
+		sock,
+		result,
+		compress,
+		&task->client_addr
+	);
+}
+
+/**
+ * Wait for an incoming request and schedule it.
+ *
+ * 1a. If the schedule is empty, we wait infinite time.
+ * 1b. If we have scheduled requests, we only wait for incoming requests
+ *     until we reach the scheduling deadline.
+ * 2a. If the incoming request was sent to a multicast destination IPv6,
+ *     choose a random delay between 0 and max_multicast_delay milliseconds.
+ * 2b. If the schedule is full, send the reply immediately.
+ * 2c. If the incoming request was sent to a unicast destination, the response
+ *     will be also sent immediately.
+ */
+static void accept_request(struct request_schedule *schedule, int sock,
+                           struct interface_delay_info *if_delay_info_list) {
+	char input[REQUEST_MAXLEN];
+	ssize_t input_bytes;
+	struct sockaddr_in6 addr;
+	char control[256];
+	struct in6_addr destaddr = {};
+	struct cmsghdr *cmsg;
+	unsigned int ifindex;
+
+	int64_t timeout = schedule_idle_time(schedule);
+	if (timeout < 0)
+		return;
+
+	// set timeout to the socket
+	struct timeval t;
+	t.tv_sec = ((uint64_t) timeout) / 1000;
+	t.tv_usec = (((uint64_t) timeout) % 1000) * 1000;
+
+	if (setsockopt(sock, SOL_SOCKET, SO_RCVTIMEO, &t, sizeof(t)) < 0)
+		perror("setsockopt failed\n");
+
+	struct iovec iv = {
+		.iov_base = input,
+		.iov_len = sizeof(input) - 1
+	};
+
+	struct msghdr mh = {
+		.msg_name = &addr,
+		.msg_namelen = sizeof(addr),
+		.msg_iov = &iv,
+		.msg_iovlen = 1,
+		.msg_control = control,
+		.msg_controllen = sizeof(control)
+	};
+
+	input_bytes = recvmsg(sock, &mh, 0);
+	update_time();
+
+	// Timeout
+	if (input_bytes < 0 && (errno == EAGAIN || errno == EWOULDBLOCK))
+		return;
+
+	if (input_bytes < 0) {
+		perror("recvmsg failed");
+		exit(EXIT_FAILURE);
+	}
+
+	// determine destination address
+	for (cmsg = CMSG_FIRSTHDR(&mh);	cmsg != NULL; cmsg = CMSG_NXTHDR(&mh, cmsg))
+	{
+		// skip other packet headers
+		if (cmsg->cmsg_level != IPPROTO_IPV6 || cmsg->cmsg_type != IPV6_PKTINFO)
+			continue;
+
+		struct in6_pktinfo *pi = (struct in6_pktinfo *) CMSG_DATA(cmsg);
+		destaddr = pi->ipi6_addr;
+		ifindex = pi->ipi6_ifindex;
+		break;
+	}
+
+	input[input_bytes] = 0;
+
+	// get the max delay
+	uint64_t max_multicast_delay = MAX_MULTICAST_DELAY_DEFAULT;
+	struct interface_delay_info *tmp = if_delay_info_list;
+	for (; tmp; tmp = tmp->next) {
+		if (tmp->ifindex == ifindex)
+			max_multicast_delay = tmp->max_multicast_delay;
```

Comment:
Also add a break here to skip the remaining loop iterations when an entry has been found.